app/src/main/java/com/github/jorgecastillo/kotlinandroid/data/HeroesRepository.kt

```diff
@@ -2,7 +2,11 @@ package com.github.jorgecastillo.kotlinandroid.data
 
 import com.github.jorgecastillo.kotlinandroid.data.datasource.remote.fetchAllHeroes
 import com.github.jorgecastillo.kotlinandroid.data.datasource.remote.fetchHeroesFromAvengerComics
+import com.karumi.marvelapiclient.model.CharacterDto
+import kategory.HK
 
-fun getHeroesWithCachePolicy() = fetchAllHeroes()
+inline fun <reified F> getHeroesWithCachePolicy(): HK<F, List<CharacterDto>> =
```

Comment:
You were not using typeclasses but just the datatypes, that is still the same before as now. If you were using Monad or others directly in your code like in this case you'd have to make them inline reified, but if you use the datatype `AsyncResult` directly then you don't have to.

This is a lang limitation. You can't defer instance resolution unless you carry the type information until the end and the only way to do that in Kotlin is through inline reified.